### ISSUE TITLE
VSCode reconnection issues for remote connections

### ISSUE THREAD

**Issue by tobespc** [eclipse/codewind#1776]
To recreate

-  Create a remote connection (can be docker desktop) from VScode
- Swap your network connection from IBM to IBM Intranet
- Wait a few seconds
- Swap network back to IBM
- connection is now lost in VScode and we fail to reauthenticate

To fix on Mac, bring up the keychain and delete the most recent codewind refresh token, then stop and start the remote connection to force a login.

**Comment by tobespc** [comment]
we are investigating this from a portal pov but could have a vscode view as well

**Comment by tetchel** [comment]
Refreshing the connection doesn't fix it?

**Comment by tobespc** [comment]
nope, even deleting the connection and recreating it doesn't work.  Might be a docker desktop issue

**Comment by markcor11** [comment]
It sounds like the IDE has cached a bad / expired access token and tries to use that when re-establishing a connection to the UI socket.  I noticed when I manually deleted the access_token and refresh_token from the keychain that they were not re-created when toggling the connection off/on from the connections list button.  (only going through the connections screen and re-saving the password did that - which then makes the connection work again).  So that suggests that the IDE is not calling the cwctl to get a new set of tokens when re-starting the connection.

**Comment by tetchel** [comment]
It makes sense to refresh the access token when the connection is refreshed. I can check that out.

**Comment by jopit** [comment]
Doesn't the ability to refresh the access token rely on the access token being valid?  If it does, and this scenario is causing the access token to be invalidated (as opposed to just being expired), then refreshing the token won't be enough, we'll need to re-authenticate.

I assume that underlying this is that when the gatekeeper sees the current access token being presented from an IP address that differs from the one it was issued on, it rejects/invalidates it.

**Comment by markcor11** [comment]
@jopit the refresh-token contains enough information for Keycloak to determine the user and works independently of any previously issued access-tokens. The refresh-token can be used instead of a username:password to request an access-token through cwctl but unlike username:passwords the refresh token has a much shorter lifespan making them more suitable for session based caching.  

Refresh tokens should be sent to `cwctl sectoken refresh` only.  This command will fetch and return a new access-token and a new refresh-token. The refresh-token should be stashed until needed, the new access-token should then be used in future API calls until such time as it expires.  Once the access-token has or is about to expire, the refresh cycle should kick in with the IDE asking cwctl for a new access-token first by calling `cwctl sectoken refresh` and if that fails (because the refresh might also have expired) calling the `cwctl sectoken get` which will perform a regular username:password auth request with credentials from the keychain.

Important points here are: 

access-tokens expire before refresh-tokens
refresh-tokens are only to be used to request access-tokens
Codewind API requests should include access-tokens not refresh-tokens

For Socket IO connections the behaviour is a little different.  

When the IDE connects to the Codewind UISocket it must emit an authentication event with a access-token payload.  The access-token must be valid (not expired) and must have originated from a Keycloak service trusted by the Codewind deployment. Codewind will check the access-token and validate it using the realm and codewind-client public key stored in Keycloak. Assuming the access-token is good, the socket is considered authenticated and streaming of socket messages can proceed until such time as the socket connection drops.   

Important point here is : socket IO connections remain authenticated until the socket drops and beyond the lifespan of the access_token. 

When the socket drops (which could be hours after the original connection) there is a very high chance the access-token originally used would have now expired. The socket IO protocol will not consume a refresh token.  (refreshing a token is a cwctl task) and for that reason, a new access token should be requested via cwctl and presented in the follow-up socket reconnection attempts. 

The IDE still has a choice of using the `cwctl sectoken refresh` or `cwctl sectoken get` but must should call cwctl to get a new access-token either way. 

You might ask,  what's the point of using `cwctl sectoken refresh` first (which might fail due to an expired refresh token) when I could just use `cwctl sectoken get` only ?  The answer is session management. Refreshing an existing token re-uses the same session where as `sectoken get` will generate a whole new session in both Keycloak and in Codewind every time its called and until the single-sign-on session expires. 

So in short,  when socket connections drop (for whatever reason) please use cwctl to request a new access_token and use that new access-token when reconnecting.

**Comment by jopit** [comment]
@tetchel @eharris369 Please take a look at Mark's comment above: https://github.com/eclipse/codewind/issues/1776#issuecomment-577913258

**Comment by jopit** [comment]
Toby was saying that even restarting the IDE doesn't fix this problem -- does that mean the IDE wasn't refreshing the access token when it started up?

**Comment by tetchel** [comment]
> does that mean the IDE wasn't refreshing the access token when it started up? 

It must refresh it on startup, the access token is only stored in extension RAM

**Comment by markcor11** [comment]
Could it be that quitting vscode doesn’t actually kill all its related processes and flush memory ? Just like if you had 2 chrome browser windows opened, closed 1 of them but leaving the other one able to retain its access to shared session storage ?

**Comment by tetchel** [comment]
Each window has its own instance of the extension, and there is definitely nothing cached when you close all windows.

**Comment by tetchel** [comment]
https://github.com/eclipse/codewind/issues/1870 related?

**Comment by tetchel** [comment]
I will make this change too:

> So in short, when socket connections drop (for whatever reason) please use cwctl to request a new access_token and use that new access-token when reconnecting.

**Comment by markcor11** [comment]
@tetchel I created #1870 because it will simplify the IDE calls to cwctl in that IDEs will only need to call `sectoken get` rather than first call `sectoken refresh` (which may fail because the supplied refresh_token has expired), then have to try `sectoken get` anyway as a fallback.

We'll keep both commands available in cwctl as it assists with troubleshooting tokens, but once #1870 lands IDEs should just use `sectoken get` for authenticating.

**Comment by tetchel** [comment]
Guys, this should be fixed now. Give it a try and verify if you don't see it anymore.

**Comment by jagraj** [comment]
Our team verified this and we can not reproduce this problem any more.

**Comment by malincoln** [comment]
@tobespc do you need to verify before closing?
